src/control/client/config.go

```diff
@@ -0,0 +1,195 @@
+	SystemName    string `yaml:"name"`
+	AccessPoints  []string `yaml:"access_points"`
+	Port          int `yaml:"port"`
+	HostList      []string `yaml:"hostlist"`
+	RuntimeDir    string `yaml:"runtime_dir"`
+	HostFile      string `yaml:"host_file"`
+	Cert          string `yaml:"cert"`
+	Key           string `yaml:"key"`
+	LogFile       string `yaml:"log_file"`
+	LogFileFormat string `yaml:"log_file_format"`
+	Path          string
+	ext           External
+}
+
+// newDefaultConfiguration creates a new instance of configuration struct
+// populated with defaults.
+func newDefaultConfiguration(ext External) Configuration {
+	return Configuration{
+		SystemName:    defaultSystemName,
+		AccessPoints:  []string{"localhost"},
+		Port:          defaultPort,
+		HostList:      []string{"localhost:10001"},
+		RuntimeDir:    defaultRuntimeDir,
+		HostFile:      "",
+		Cert:          "",
+		Key:           "",
+		LogFile:       defaultLogFile,
+		LogFileFormat: "",
+		Path:          defaultConfigPath,
+		ext:           ext,
+	}
+}
+
+// ProcessConfigFile loads a configuration file from the path given,
+// or from the default location if none is provided.  It returns a populated
+// Configuration struct based upon the default values and any config file overrides.
+func ProcessConfigFile(ConfigPath string) (Configuration, error) {
+	config := NewConfiguration()
+	if ConfigPath != "" {
+		log.Debugf("Overriding default config path with: %s", ConfigPath)
+		config.Path = ConfigPath
+	}
+
+	if !filepath.IsAbs(config.Path) {
+		newPath, err := common.GetAbsInstallPath(config.Path)
+		if err != nil {
+			return config, err
+		}
+		config.Path = newPath
+	}
+
+	err := config.LoadConfig()
+	if err != nil {
+		log.Debugf("Unable to read the configuration file from: '%s'  Using default configuration.", config.Path)
+		return config, nil
+	}
+	log.Debugf("DAOS Client config read from %s", config.Path)
+
+	return config, nil
+}
+
+// LoadConfig reads the configuration file specified by Configuration.Path
+// and parses it.  Parsed values override any default values.
+func (c *Configuration) LoadConfig() error {
+	bytes, err := ioutil.ReadFile(c.Path)
+	if err != nil {
+		return err
+	}
+
+	if err = c.parse(bytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ApplyAgentCmdLineOverrides will overwrite Configuration values with any non empty
+// data provided, usually from the commandline.
+func (c *Configuration) ApplyAgentCmdLineOverrides(RuntimeDir string, LogFile string) error {
+
+	if RuntimeDir != "" {
+		log.Debugf("Overriding socket path from config file with %s", RuntimeDir)
+		c.RuntimeDir = RuntimeDir
+	}
+
+	if LogFile != "" {
+		log.Debugf("Overriding LogFile path from config file with %s", LogFile)
+		c.LogFile = LogFile
+	}
+
+	return nil
+}
+
+// ApplyDMGCmdLineOverrides will overwrite Configuration values with any non empty
+// data provided, usually from the commandline.
+func (c *Configuration) ApplyDMGCmdLineOverrides(Hostlist string) error {
```

Comment:
Maybe not that important in the short term, but just an organizational suggestion - these DMG and agent specific overrides seem like they belong with the DMG and agent code, respectively, rather than in the shared configuration object.

Reply:
Here's a spot where I got myself into trouble with multiple branches.  In an earlier version, Tom asked me to remove the code similar to this that was potentially cluttering up the Agent and DMG.  He wanted it pushed down into the client code.  I can see both sides and am OK with either.  I'd like to know what Tom thinks at this point, if we can get consensus one way or the other.

Reply:
I've just repeated Kris' comment above. method should only live in here if they can be made generic. sorry for any confusion Joel.

Reply:
I will add these overrides to the DMG and agent code.